### ISSUE TITLE
Automatic device transfer converts namedtuples into regular tuples

### ISSUE THREAD

**Issue by lakshmi-speak** [Lightning-AI/utilities#199]
## 🐛 Bug

In the latest version of the utilities library, automatic device transfer converts namedtuples into regular tuples, causing loss of attribute access provided by the namedtuple.

### To Reproduce

```py
from collections import namedtuple
import torch
from pytorch_lightning import LightningModule, Trainer

Batch = namedtuple("Batch", ["features", "targets"])

class SimpleModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(5, 1)
        
    def forward(self, batch):
        print(type(batch))
        print(batch.features)
        return self.linear(batch.features)

    def training_step(self, batch, batch_idx):
        return self(batch)

    def train_dataloader(self):
        features = torch.randn(10, 5)
        targets = torch.randint(0, 1, (10,))
        dataset = [Batch(features[i], targets[i]) for i in range(10)]
        return torch.utils.data.DataLoader(dataset, batch_size=1)
    
    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)

model = SimpleModule()
trainer = Trainer(fast_dev_run=True)
trainer.fit(model)
```
This script fails with attribute error `AttributeError: 'tuple' object has no attribute 'features'`, when the library version is 0.10.0 and runs without errors on 0.9.0. The Pytorch lightning version was unchanged, 2.1.2 for both.

### Additional context

<details>
  <summary><b>Environment details</b></summary>

- PyTorch Version (e.g., 1.0): 2.1.1+cu121
- OS (e.g., Linux): Ubuntu
- How you installed PyTorch (`conda`, `pip`, source): pip 
- Python version: 3.10.13
- CUDA/cuDNN version: 12.2

This happened for both single GPU and multi GPU run with DDP. This was a bit of work to track down since the pinned pytorch_lightning version was unchanged.
This is the exact same as a previous lightning [issue](https://github.com/Lightning-AI/lightning/issues/1588), although that one seems to have been due to PyTorch?

**Comment by carmocca** [comment]
Hi @lakshmi-speak. Sorry for this.

I think it's best that we revert https://github.com/Lightning-AI/utilities/pull/160.

Would you like to open a PR doing that and adding your namedtuple as a test to https://github.com/Lightning-AI/utilities/blob/main/tests/unittests/core/test_apply_func.py?

**Comment by Borda** [comment]
@carmocca was this resolved in #210 too?
